sql/core/src/test/scala/org/apache/spark/sql/sources/v2/TestInMemoryTableCatalog.scala

```diff
@@ -196,7 +201,105 @@ private class InMemoryTable(
   }
 }
 
-private class BufferedRows extends WriterCommitMessage with InputPartition with Serializable {
+object TestInMemoryTableCatalog {
+  val SIMULATE_FAILED_WRITE_OPTION = "spark.sql.test.simulateFailedWrite"
+  val SIMULATE_FAILED_CREATE_PROPERTY = "spark.sql.test.simulateFailedCreate"
+
+  def maybeSimulateFailedTableCreation(tableProperties: util.Map[String, String]): Unit = {
+    if ("true".equalsIgnoreCase(
+      tableProperties.get(TestInMemoryTableCatalog.SIMULATE_FAILED_CREATE_PROPERTY))) {
+      throw new IllegalStateException("Manual create table failure.")
+    }
+  }
+
+  def maybeSimulateFailedTableWrite(tableOptions: CaseInsensitiveStringMap): Unit = {
+    if (tableOptions.getBoolean(
+      TestInMemoryTableCatalog.SIMULATE_FAILED_WRITE_OPTION, false)) {
+      throw new IllegalStateException("Manual write to table failure.")
+    }
+  }
+}
+
+class TestStagingInMemoryCatalog
+  extends TestInMemoryTableCatalog with StagingTableCatalog {
+
+  override def stageCreate(
+      ident: Identifier,
+      schema: StructType,
+      partitions: Array[Transform],
+      properties: util.Map[String, String]): StagedTable = {
+    newStagedTable(ident, schema, partitions, properties, replaceIfExists = false)
+  }
+
+  override def stageReplace(
+      ident: Identifier,
+      schema: StructType,
+      partitions: Array[Transform],
+      properties: util.Map[String, String]): StagedTable = {
+    newStagedTable(ident, schema, partitions, properties, replaceIfExists = true)
+  }
+
+  private def newStagedTable(
+      ident: Identifier,
+      schema: StructType,
+      partitions: Array[Transform],
+      properties: util.Map[String, String],
+      replaceIfExists: Boolean): StagedTable = {
+    import CatalogV2Implicits.IdentifierHelper
+    if (partitions.nonEmpty) {
+      throw new UnsupportedOperationException(
+        s"Catalog $name: Partitioned tables are not supported")
+    }
+
+    TestInMemoryTableCatalog.maybeSimulateFailedTableCreation(properties)
+
+    new TestStagedTable(
+      ident,
+      new InMemoryTable(s"$name.${ident.quoted}", schema, properties),
+      replaceIfExists)
+  }
+
+  private class TestStagedTable(
+      ident: Identifier,
+      delegateTable: InMemoryTable,
+      replaceIfExists: Boolean)
+    extends StagedTable with SupportsWrite with SupportsRead {
+
+    override def commitStagedChanges(): Unit = {
+      if (droppedTables.contains(ident)) {
```

Comment:
it's weird to record all the dropped tables in the history. I think a simple version is
```
if (replaceIfExists) {
  if (!tables.containsKey(ident)) {
      throw new RuntimeException("table already dropped")
  }
  tables.put(ident, delegateTable)
}
```

Reply:
That doesn't work because the implementation of `stageCreate` doesn't actually put the table in the `tables` map at all. So you can't necessarily say the table was dropped just because the table is not in the `tables` map.

Reply:
It's unclear to me if this is the correct behavior - if something dropped the table from underneath this, the subsequent commit of the replace or atomic-create operation should have the final say, right?

Reply:
Think about a REPLACE TABLE and DROP TABLE happen at the same time. It doesn't matter which one gets executed first, but the final result match be reachable by one certain execution order.

If REPLACE TABLE executes first, then there should be no table at the end as it's dropped.
If DROP TABLE executes first, then REPLACE TABLE should fail and there is still no table at the end.

Reply:
BTW I think my proposal works for REPLACE TABLE right? `stageCreate` is for CTAS and I think your current code already works

Reply:
I thought about this a bit more and chatted with @rdblue and I realized the confusion is that the staging catalog API doesn't even support passing through the `orCreate` flag to the catalog. I think we need to pass this information along to the catalog, otherwise the catalog won't know that the user wanted `CREATE OR REPLACE` semantics.

I'm more inclined to add an extra method to `StagingTableCatalog` called `stageCreateOrReplace`, in addition to the other methods we have here already. Then the behavior of `commitStagedChanges` depends on whether or not the table was instantiated via `stageCreateOrReplace` vs. `replace` vs. `create`.